### PR TITLE
Generating bindings for WinTrust

### PR DESCRIPTION
**Pull request by nathan-alden-sr**
Generated bindings for `WinTrust.h`.

### CODE REVIEW

File: generation/um/WinTrust/generate.rsp [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+--file
+um-WinTrust.h
+--headerFile
+header.txt
+--include-directory
+C:/Program Files (x86)/Windows Kits/10/Include/10.0.18362.0/shared
+--include-directory
+C:/Program Files (x86)/Windows Kits/10/Include/10.0.18362.0/um
+--libraryPath
+Wintrust
```
Comment (tannergooding) [inline generation/um/WinTrust/generate.rsp:14]: nit: This should be lowercase to match the actual filename `wintrust.dll`

File: sources/Interop/WinTrust/WinCodec.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// Copyright © Tanner Gooding and Contributors. Licensed under the MIT License (MIT). See License.md in the repository root for more information.
+
+namespace TerraFX.Interop
```
Comment (tannergooding) [inline sources/Interop/WinTrust/WinCodec.cs:3]: Don't think you meant to include this file 😄

File: sources/Interop/WinTrust/um/wintrust/CAT_MEMBERINFO2.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+// Ported from um/WinTrust.h in the Windows SDK for Windows 10.0.18362.0
+// Original source is Copyright © Microsoft. All rights reserved.
+
+namespace TerraFX.Interop
+{
+    public partial struct CAT_MEMBERINFO2
+    {
+        [NativeTypeName("GUID")]
+        public _GUID SubjectGuid;
```
Comment (tannergooding) [inline sources/Interop/WinTrust/um/wintrust/CAT_MEMBERINFO2.cs:11]: `_GUID` should be remapped to `Guid` (which will automatically add the appropriate `using System;`, etc)

File: sources/Interop/WinTrust/um/wintrust/CAT_NAMEVALUE.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    public unsafe partial struct CAT_NAMEVALUE
+    {
+        [NativeTypeName("LPWSTR")]
+        public ushort* pwszTag;
+
+        [NativeTypeName("DWORD")]
+        public uint fdwFlags;
+
+        [NativeTypeName("CRYPT_DATA_BLOB")]
+        public _CRYPTOAPI_BLOB Value;
```
Comment (tannergooding) [inline sources/Interop/WinTrust/um/wintrust/CAT_NAMEVALUE.cs:17]: `_CRYPTOAPI_BLOB` should be remapped. I think it also needs to be defined.

Comment (tannergooding) [reply sources/Interop/WinTrust/um/wintrust/CAT_NAMEVALUE.cs:17]: Looks like its defined in `um\wincrypt.h`. It looks like it belongs in `Advapi32.dll` for the most part, based on the methods in the header...

File: sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_CERT.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+namespace TerraFX.Interop
+{
+    public unsafe partial struct CRYPT_PROVIDER_CERT
+    {
+        [NativeTypeName("DWORD")]
+        public uint cbStruct;
+
+        [NativeTypeName("PCCERT_CONTEXT")]
+        public _CERT_CONTEXT* pCert;
```
Comment (tannergooding) [inline sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_CERT.cs:14]: Should be remapped and defined also from wincrypt.

Comment (tannergooding) [reply sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_CERT.cs:14]: Same with `_CTL_CONTEXT` and `_CERT_CHAIN_ELEMENT` below

File: sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_DATA.cs [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        public uint cbStruct;
+
+        [NativeTypeName("WINTRUST_DATA *")]
+        public WINTRUST_DATA* pWintrustData;
+
+        [NativeTypeName("BOOL")]
+        public int fOpenedFile;
+
+        [NativeTypeName("HWND")]
+        public HWND__* hWndParent;
```
Comment (tannergooding) [inline sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_DATA.cs:22]: Remap using `HWND__*=IntPtr` for these ones.
Also `_GUID` below

File: sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_DEFUSAGE.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace TerraFX.Interop
+{
+    public unsafe partial struct CRYPT_PROVIDER_DEFUSAGE
+    {
+        [NativeTypeName("DWORD")]
+        public uint cbStruct;
+
+        [NativeTypeName("GUID")]
+        public _GUID gActionID;
```
Comment (tannergooding) [inline sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_DEFUSAGE.cs:14]: Needs remapping

File: sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_PRIVDATA.cs [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace TerraFX.Interop
+{
+    public unsafe partial struct CRYPT_PROVIDER_PRIVDATA
+    {
+        [NativeTypeName("DWORD")]
+        public uint cbStruct;
+
+        [NativeTypeName("GUID")]
+        public _GUID gProviderID;
```
Comment (tannergooding) [inline sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_PRIVDATA.cs:14]: Needs remapping

File: sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_FUNCTIONS.cs [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        public IntPtr* pfnCertificateTrust;
+
+        [NativeTypeName("PFN_PROVIDER_FINALPOLICY_CALL")]
+        public IntPtr* pfnFinalPolicy;
+
+        [NativeTypeName("PFN_PROVIDER_CERTCHKPOLICY_CALL")]
+        public IntPtr* pfnCertCheckPolicy;
+
+        [NativeTypeName("PFN_PROVIDER_TESTFINALPOLICY_CALL")]
+        public IntPtr* pfnTestFinalPolicy;
```
Comment (tannergooding) [inline sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_FUNCTIONS.cs:52]: I think I borked this when I added the experimental function pointer support...

Comment (tannergooding) [reply sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_FUNCTIONS.cs:52]: It is supposed to be just `IntPtr`... I might just leave for now and fix when I regenerate the full bindings for 20H1

File: sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_REGDEFUSAGE.cs [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace TerraFX.Interop
+{
+    public unsafe partial struct CRYPT_PROVIDER_REGDEFUSAGE
+    {
+        [NativeTypeName("DWORD")]
+        public uint cbStruct;
+
+        [NativeTypeName("GUID *")]
+        public _GUID* pgActionID;
```
Comment (tannergooding) [inline sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_REGDEFUSAGE.cs]: Needs remapping

File: generation/um/WinTrust/generate.rsp (hunk 1 above)
Comment (nathan-alden-sr) [reply generation/um/WinTrust/generate.rsp]: Fixed.

File: sources/Interop/WinTrust/WinCodec.cs (hunk 2 above)
Comment (nathan-alden-sr) [reply sources/Interop/WinTrust/WinCodec.cs]: Fixed.

File: sources/Interop/WinTrust/um/wintrust/CAT_MEMBERINFO2.cs (hunk 3 above)
Comment (nathan-alden-sr) [reply sources/Interop/WinTrust/um/wintrust/CAT_MEMBERINFO2.cs]: Fixed.

File: sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_SGNR.cs [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace TerraFX.Interop
+{
+    public unsafe partial struct CRYPT_PROVIDER_SGNR
+    {
+        [NativeTypeName("DWORD")]
+        public uint cbStruct;
+
+        [NativeTypeName("FILETIME")]
+        public _FILETIME sftVerifyAsOf;
```
Comment (tannergooding) [inline sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_SGNR.cs:14]: Needs remapping, should be already defined in `TerraFX.Interop.Windows`

File: sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_DATA.cs (hunk 6 above)
Comment (nathan-alden-sr) [reply sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_DATA.cs]: Fixed.

File: sources/Interop/WinTrust/um/wintrust/PROVDATA_SIP.cs [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    public unsafe partial struct PROVDATA_SIP
+    {
+        [NativeTypeName("DWORD")]
+        public uint cbStruct;
+
+        [NativeTypeName("GUID")]
+        public _GUID gSubject;
+
+        [NativeTypeName("struct SIP_DISPATCH_INFO_ *")]
+        public SIP_DISPATCH_INFO_* pSip;
```
Comment (tannergooding) [inline sources/Interop/WinTrust/um/wintrust/PROVDATA_SIP.cs:19]: These ones are properly defined in `sipbase.h` which looks like it is part of `Crypt32.dll`

File: sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_DEFUSAGE.cs (hunk 7 above)
Comment (nathan-alden-sr) [reply sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_DEFUSAGE.cs]: Fixed.

File: sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_PRIVDATA.cs (hunk 8 above)
Comment (nathan-alden-sr) [reply sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_PRIVDATA.cs]: Fixed.

File: sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_REGDEFUSAGE.cs (hunk 10 above)
Comment (nathan-alden-sr) [reply sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_REGDEFUSAGE.cs]: Fixed.

File: sources/Interop/WinTrust/um/wintrust/PROVDATA_SIP.cs [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+        [NativeTypeName("struct SIP_SUBJECTINFO_ *")]
+        public SIP_SUBJECTINFO_* psSipSubjectInfo;
+
+        [NativeTypeName("struct SIP_SUBJECTINFO_ *")]
+        public SIP_SUBJECTINFO_* psSipCATSubjectInfo;
+
+        [NativeTypeName("struct SIP_INDIRECT_DATA_ *")]
+        public SIP_INDIRECT_DATA_* psIndirectData;
+
+        public partial struct SIP_DISPATCH_INFO_
```
Comment (tannergooding) [inline sources/Interop/WinTrust/um/wintrust/PROVDATA_SIP.cs:33]: You'll need to exclude these names to get the partial structs dropped in favor of the ones from `sipbase.h`

File: sources/Interop/WinTrust/um/wintrust/CAT_NAMEVALUE.cs (hunk 4 above)
Comment (nathan-alden-sr) [reply sources/Interop/WinTrust/um/wintrust/CAT_NAMEVALUE.cs]: Fixed.

File: sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_CERT.cs (hunk 5 above)
Comment (nathan-alden-sr) [reply sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_CERT.cs]: Fixed.

File: sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_FUNCTIONS.cs (hunk 9 above)
Comment (nathan-alden-sr) [reply sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_FUNCTIONS.cs:52]: Not sure what I can do to fix this.

File: sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_SGNR.cs (hunk 11 above)
Comment (nathan-alden-sr) [reply sources/Interop/WinTrust/um/wintrust/CRYPT_PROVIDER_SGNR.cs]: Fixed.

File: sources/Interop/WinTrust/um/wintrust/PROVDATA_SIP.cs (hunk 13 above)
Comment (nathan-alden-sr) [reply sources/Interop/WinTrust/um/wintrust/PROVDATA_SIP.cs]: Fixed.

File: sources/Interop/WinTrust/um/wintrust/PROVDATA_SIP.cs (hunk 12 above)
Comment (nathan-alden-sr) [reply sources/Interop/WinTrust/um/wintrust/PROVDATA_SIP.cs]: Fixed.

File: sources/Interop/Advapi32/um/wincrypt/Advapi32.Manual.cs [hunk 14]
```diff
@@ -0,0 +1,5951 @@
...
+
+        public const int CERT_SYSTEM_STORE_CURRENT_USER_GROUP_POLICY = CERT_SYSTEM_STORE_CURRENT_USER_GROUP_POLICY_ID << CERT_SYSTEM_STORE_LOCATION_SHIFT;
+
+        public const int CERT_SYSTEM_STORE_LOCAL_MACHINE_GROUP_POLICY = CERT_SYSTEM_STORE_LOCAL_MACHINE_GROUP_POLICY_ID << CERT_SYSTEM_STORE_LOCATION_SHIFT;
+
+        public const int CERT_SYSTEM_STORE_LOCAL_MACHINE_ENTERPRISE = CERT_SYSTEM_STORE_LOCAL_MACHINE_ENTERPRISE_ID << CERT_SYSTEM_STORE_LOCATION_SHIFT;
+
+        public const int CERT_SYSTEM_STORE_LOCAL_MACHINE_WCOS = CERT_SYSTEM_STORE_LOCAL_MACHINE_WCOS_ID << CERT_SYSTEM_STORE_LOCATION_SHIFT;
+
+        public const string CERT_GROUP_POLICY_SYSTEM_STORE_REGPATH =
```
Comment (tannergooding) [inline sources/Interop/Advapi32/um/wincrypt/Advapi32.Manual.cs:5919]: These are (mostly) just macro functions that weren't ported, right?

File: sources/Interop/Advapi32/um/wincrypt/Advapi32.cs [hunk 15]
```diff
@@ -0,0 +1,1082 @@
...
+
+namespace TerraFX.Interop
+{
+    public static unsafe partial class Advapi32
+    {
+        private const string LibraryPath = "advapi32";
+
+        [DllImport(LibraryPath, CallingConvention = CallingConvention.Winapi, EntryPoint = "CryptAcquireContextA", ExactSpelling = true)]
+        [return: NativeTypeName("BOOL")]
+        public static extern int CryptAcquireContextA([NativeTypeName("HCRYPTPROV *")] uint* phProv, [NativeTypeName("LPCSTR")] sbyte* szContainer, [NativeTypeName("LPCSTR")] sbyte* szProvider, [NativeTypeName("DWORD")] uint dwProvType, [NativeTypeName("DWORD")] uint dwFlags);
```
Comment (tannergooding) [inline sources/Interop/Advapi32/um/wincrypt/Advapi32.cs:17]: Several of these, like `HCRYPTPROV` are handle types that don't go through the same define as other handles like `HWND`

You should be able to get most of them by remapping `ULONG_PTR=UIntPtr` and `LONG_PTR=IntPtr`

File: sources/Interop/Advapi32/um/wincrypt/Advapi32.Manual.cs (hunk 14 above)
Comment (nathan-alden-sr) [reply sources/Interop/Advapi32/um/wincrypt/Advapi32.Manual.cs:5919]: I'm a bit confused about how your question relates to the code you highlighted. There are many `#define`s in `wincrypt.h` that are basically simple math expressions (like the one you highlighted). However, I did not port any macro functions at all throughout this PR (i.e., anything with `#define F(X)`.

Comment (nathan-alden-sr) [reply sources/Interop/Advapi32/um/wincrypt/Advapi32.Manual.cs:5919]: The code you highlighted is defined thusly:
```c
#define CERT_SYSTEM_STORE_LOCAL_MACHINE_WCOS  \
    (CERT_SYSTEM_STORE_LOCAL_MACHINE_WCOS_ID << \
        CERT_SYSTEM_STORE_LOCATION_SHIFT)
```

File: sources/Interop/Advapi32/um/wincrypt/Advapi32.cs (hunk 15 above)
Comment (nathan-alden-sr) [reply sources/Interop/Advapi32/um/wincrypt/Advapi32.cs]: Fixed.

From my research it seems like only `ULONG_PTR` was required and only for `wincrypt.h`.

File: sources/Interop/Advapi32/um/wincrypt/Advapi32.Manual.cs (hunk 14 above)
Comment (tannergooding) [reply sources/Interop/Advapi32/um/wincrypt/Advapi32.Manual.cs:5919]: For whatever reason, the "Conversation" tab puts this on the wrong line.

If you goto the "Files" tab, it is on L5919: https://github.com/terrafx/terrafx.interop.windows/pull/37/files/435c24606296062bd557b116d453c517e1e042cb#diff-d03360204625b1a7168b85116fe6c24bR5919

Comment (nathan-alden-sr) [reply sources/Interop/Advapi32/um/wincrypt/Advapi32.Manual.cs:5919]: Ah, yes, those are all the remaining `#define`s. The PowerShell script creates comments for every `#define` that couldn't be mapped automatically. I then go through the list and convert any that I feel like I can manage. Most of the leftovers are "functions."

**Comment by tannergooding** [comment]
Thanks for the contribution!

**Pull request merged**